### PR TITLE
No offload for WB store ops

### PR DESCRIPTION
**Pull request by ahmetmircik**
Seen during investigations of https://github.com/hazelcast/hazelcast/issues/22601

**Issue:**
Stores are not blocking when write-behind is in use therefore they don't need any offload. (Loads are different and even write-behind is in use they are blocking so offload is needed for them). This offload, handing map-store api to a different thread, creates a perf penalty in write behind case.

**Fix:**
This PR creates two sub-types for offloadable Step, as load and store types, only `load` types are offloaded when write-behind is enabled and `store` types are processed by partition threads.

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/StepRunner.java [hunk 1]
```diff
@@ -217,8 +216,9 @@ private void offloadRun(Runnable step,
             operationExecutor.execute(offload);
         } else {
             Offloadable offloadableStep = (Offloadable) step;
-            executorName = offloadableStep.getExecutorName();
-            executionService.getExecutor(executorName).execute(offload);
+            currentExecutorName = offloadableStep.getExecutorName();
+            // TODO what if there is RejectedExecutionException
```
Comment (frant-hartm) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/StepRunner.java:220]: Let's either resolve the TODO or create an issue.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/PutOpSteps.java [hunk 2]
```diff
@@ -87,45 +88,47 @@ public void runStep(State state) {
...
+                if (staticParams.isPutIfAbsent()) {
+                    Record record = recordStore.getRecord(state.getKey());
+                    if (record == null && state.getOldValue() != null) {
+                        record = ((DefaultRecordStore) recordStore).onLoadRecord(state.getKey(),
+                                state.getOldValue(), false, state.getCallerAddress());
+                    }
+
+                    if (record != null) {
+                        state.setOldValue(record.getValue());
+                        state.setStopExecution(true);
```
Comment (frant-hartm) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/PutOpSteps.java:101]: Is this just refactoring, or was this change needed to fix something?

**Review by frant-hartm — APPROVED** [review]

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/StepRunner.java (hunk 1 above)
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/StepRunner.java:220]: it is here https://github.com/hazelcast/hazelcast/pull/22684

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/PutOpSteps.java (hunk 2 above)
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/PutOpSteps.java:101]: no fix, it is about aligning it with no-offload case.

**Review by vbekiaris — COMMENTED** [review]
Good job. We should add assertions that store methods are invoked from partition operation thread in `MapStoreWriteBehindTest` to avoid a future regression.
Also, is there a PR for `master`?

**Comment by ahmetmircik** [comment]
@vbekiaris add assertions in IMapOpsStep. Upon merge, will send a master counterpart.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/IMapOpStep.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        return false;
+    }
+
+    default boolean isWriteBehind(State state) {
+        return state.getRecordStore().getMapDataStore()
+                instanceof WriteBehindStore;
+    }
+
+    default void assertWBStoreRunsOnPartitionThread(State state) {
+        if (isStoreStep() && isWriteBehind(state)) {
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/IMapOpStep.java:66]: Let's rewrite this so we avoid branching altogether when running with assertions disabled:
```
assert (isStoreStep() && isWriteBehind(state) ? ThreadUtil.isRunningOnPartitionThread() : true);
```

Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/IMapOpStep.java]: makes sense, done.

**Review by vbekiaris — APPROVED** [review]
